// SPDX-FileCopyrightText: 2023 Pollux authors <https://github.com/polluxio/pollux-payload/blob/main/AUTHORS>
// SPDX-License-Identifier: Apache-2.0

syntax = "proto3";

option go_package = "pollux/private/grpc";

package pollux;

import "pollux.proto";

//Pollux server from Payload side
service PolluxPayload {
  rpc Transmit(PolluxMessage) returns (PolluxMessageResponse) {}
  rpc Start(PayloadStartMessage) returns (PolluxControlResponse) {}
  rpc Iterate(PayloadIterateMessage) returns (PolluxControlResponse) {}
  rpc Terminate(PayloadTerminateMessage) returns (EmptyResponse) {}
}

//Pollux Payload server from Zebulon side
service ZebulonPayload {
  rpc PayloadReady(PayloadReadyMessage) returns (PolluxStandardResponse) {}
  rpc PayloadLoopReadyForNextIteration(PayloadLoopMessage) returns (PolluxStandardResponse) {}
  rpc PayloadLoopEnd(PayloadLoopMessage) returns (PolluxStandardResponse) {}
  rpc PayloadInactive(PayloadInactiveMessage) returns (PolluxStandardResponse) {}
  rpc Transmit(stream PolluxMessage) returns (PolluxMessageResponse) {}
  rpc PolluxReport(PolluxReportMessage) returns (PolluxStandardResponse) {}
}

message PayloadReadyMessage {
  string info = 1;
  uint32 port = 2;
  pollux.PolluxVersion version = 3;
}

message PayloadLoopMessage {
  uint32 iteration = 1;
}

message PayloadInactiveMessage {
  string info = 1;
  uint64 since = 2;
}

message PayloadStartMessage {
  string info = 1;
  pollux.PolluxControl control = 2;
}

message PayloadIterateMessage {
  string info = 1;
  uint32 iteration = 2;
}

message PayloadTerminateMessage {
  string info = 1;
}
